### PR TITLE
UI Automation in Windows Console: make default in FORMATTED consoles (Windows 11 Sun Valley 2)

### LINKED ISSUE #8653: WHen NVDA is started from source with python.exe it is not possible to interact with any console window.
### Steps to reproduce:
1. On system with all NVDA dependencies installed clone NVDA git repository and run scons source.
2. move to the source directory and type python nvda.pyw
3. Try to read a console window in which NVDA is started.
4. Open another cmd and try to interact with it.
### Actual behavior:
Screen reviev keys didn't work and the following is written in the log each time when the console is focused.
ERROR - NVDAObjects.behaviors.LiveText._monitor (14:54:56.263):
Error getting initial lines
Traceback (most recent call last):
  File "NVDAObjects\behaviors.py", line 260, in _monitor
    oldLines = self._getTextLines()
  File "NVDAObjects\window\winConsole.py", line 49, in _getTextLines
    return winConsoleHandler.getConsoleVisibleLines()
  File "winConsoleHandler.py", line 121, in getConsoleVisibleLines
    consoleScreenBufferInfo=wincon.GetConsoleScreenBufferInfo(consoleOutputHandle)
  File "wincon.py", line 70, in GetConsoleScreenBufferInfo
    raise WinError()
WindowsError: [Error 6] Nieprawidłowe dojście. which in polish mean something like the handle is invalid.

### Expected behavior:
NVDA should be able to read a console window.
#### NVDA Installed/portable/running from source:
Running from source
#### NVDA version:
master commit ec3eeee941f6b47f6ebb8ead9dc75bdc3efddfa1 (HEAD -> master, nvaccess/master
#### Windows version:
Windows 7 x64
#### Name and version of other software in use when reproducing the issue:
Python 2.7.15, Visual Studio community 2017
#### Other information about your system:
None
#### Does the issue still occur after restarting your PC?
Yes
#### Have you tried any other versions of NVDA?
No

### LINKED ISSUE #9867: Could not attach console (windows error 6) when running cmd.exe
Note: the below only occurs if the advanced "Use
UI Automation to access the windows console when available" option is turned off. In
other words: the current default condition.

### Steps to reproduce:

Run a recent NVDA alpha or beta.
Perform Windows+r, "cmd", enter.

### Actual behavior:

A windows error sound occurs on launch.

The console appears to act normally after that, although when doing a "cls" NVDA is silent (does not read the prompt), until enter is pressed a second time, though this may be unrelated and happens in older versions.

Various warnings, and a "could not attach console: windows error (6, 'the handle is invalid')" are logged.

### Expected behavior:

No windows error sound, no logged warnings or errors.

#### NVDA installed/portable/running from source:

Installed

#### NVDA version:

beta-17929

#### Windows version:

Windows 10 Home 1903, 64 bit

#### Name and version of other software in use when reproducing the issue:

Windows command shell (cmd.exe)

#### Other information about your system:

AMD

#### Does the issue still occur after restarting your PC?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

A few recent betas have done this.
Edit: also occurs on an installed version of alpha-17928,9dfbd2ee (7/3/2019).
Doesn't happen with 2019.1.1, although the (likely unrelated) cls problem does.

[nvda-cmd-errors.txt](https://github.com/nvaccess/nvda/files/3356708/nvda-cmd-errors.txt)

### LINKED ISSUE #11172: Efficiently reviewing the output buffer in terminals
CC: @codeofdusk

### Is your feature request related to a problem? Please describe.

I couldn't figure out how to do the following:

1. Efficiently review text that has scrolled offscreen.
2. Efficiently copy parts of the buffer to the clipboard that have scrolled offscreen
3. Find occurances of a search term of interest, which is really useful in many situations. Both the legacy terminal and the new Windows Terminal have a find function which visually highlights matches, but there's no way to move to it with NVDA.
4. Probably slightly out of scope: dealing with text-based tables. For example, a command line disassembler usually has the memory address in the first column, which you are subjected to hearing on every line when using NVDA even though you're actually interested in the actual CPU instructions. Many useful commands also output statistics in tables which is hard to read.

### Describe the solution you'd like

These are just some ideas for now, I'd love to hear how others have solved these issues.

1. Logging all output to a virtual document: using a key to switch to it, maybe nvda + space, you would then be able to review past output up to e.g, the past 1000 lines. I'm not sure what the performance implications of this are, but it might allow for some possibilities in the future such as turning text-based tables into a table that NVDA recognizes, or turning each prompt line into a heading for faster navigation.
2. Make use of the hotkeys in the terminal: [conhost has the mark mode for this](https://docs.microsoft.com/en-us/previous-versions/orphan-topics/ws.11/mt427362(v=ws.11)?redirectedfrom=MSDN) but there are some oddities when trying to use it in practice. This also does not work with the legacy terminal in WSL and is also not supported in Windows Terminal, see microsoft/terminal#5948.
3. Use a paging utility

### Describe alternatives you've considered

I have therefore been forced to copy all text and paste it into a blank notepad document just to do some of these basic tasks.

### LINKED ISSUE #11554: Passwords spoken in WSL
### Steps to reproduce:

1. If not already enabled and installed, enable WSL, and install and setup a distro (I am using Ubuntu 20.04 on WSL2, but I observed this issue on WSL1).
2. Make sure "Speak typed characters" is enabled in NVDA.
3. Within the WSL distro, try to do something that requires a password (EG run a command via `sudo`).

### Actual behavior:

The password is spoken by NVDA, even though it is not displayed on screen.

### Expected behavior:

The password should not be spoken, or, a masking symbol (such as "*") should be spoken instead.

#### NVDA installed/portable/running from source:

Installed

#### NVDA version:

2020.2

#### Windows version:

Microsoft Windows [Version 10.0.19041.450]

#### Name and version of other software in use when reproducing the issue:

Windows Subsystem for Linux with Ubuntu 20.04.

#### Does the issue still occur after restarting your computer?

Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

I believe this issue occurs in prior versions of NVDA, but am going from memory.

#### If addons are disabled, is your problem still occurring?

Yes.

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?

Yes. Same result before and after.

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
Supersedes #9771 and #10716. Closes #1682. Closes #8653. Closes #9867. Closes #11172. Closes #11554.

### Summary of the issue:
Microsoft has significantly improved performance and reliability of UIA console:
* microsoft/terminal#4018 is an almost complete rewrite of the UIA code which makes the console's UIA implementation more closely align with the API specification.
* microsoft/terminal#10886, microsoft/terminal#10925, and microsoft/terminal#11253 form a robust testing methodology for the UIA implementation, including bug fixes in response to newly added tests based on Word's UIA implementation.
* microsoft/terminal#11122 removes the thousands of empty lines at the end of the console buffer, significantly improving performance and stability. Since all console text ranges are now within the text buffer's bounds, it is no longer possible for console to crash due to the manipulation by UIA clients of an out-of-bounds text range.
* Countless other accessibility-related PRs too numerous to list here.

We should enable UIA support on new Windows Console builds by default for performance improvement and controllable password suppression.

### Description of how this pull request fixes the issue:
This PR:
* Exposes all three options for the UIA console feature flag in the UI (replaces the UIA check box with a combo box).
* Adds a runtime check to test if `apiLevel >= FORMATTED`, and use UIA in this case when the user preference is auto. This is the case on Windows 11 Sun Valley 2 (SV2) available now in beta and set for release in the second half of 2022.

### Testing performed:
Tested the `IMPROVED` and `FORMATTED` console implementations as my primary command console for over a year (since February 2020) and worked closely with the Windows Console team (especially @carlos-zamora) to verify stability and report/fix issues. Sent test builds of the new console to several others to verify stability and functionality.

* When UIA console is set to auto in advanced settings, UIA is used in the Windows 11 SV2 console and legacy in all older versions.
* When UIA console is set to UIA in advanced settings, UIA is always used if available (console has a UIA provider and legacy mode in properties is unticked).
* When UIA console is set to legacy in advanced settings, legacy is always used.

### Known issues with pull request:
In the event that this PR must be reverted, please do a `git revert` or the GitHub revert button to roll it back completely (I'm happy to help with conflict resolution should it be necessary). In other words, please do not change the default value of the `winConsoleImplementation` option in the config spec in an attempt to back out this change similar to #10682, as the "auto" option will then not match default behaviour as expected.

### Change log entry:
== New Features ==
- In the Windows Console Host used by Command Prompt, PowerShell, and the Windows Subsystem for Linux on Windows 11 version 22H2 (Sun Valley 2) and later:
    - Vastly improved performance and stability. (#10964)
    - When pressing ctrl+f to find text, the review cursor position is updated to follow the found term. (#11172)
    - Reporting of typed text that does not appear onscreen (such as passwords) is disabled by default. It can be re-enabled in NVDA's advanced settings panel. (#11554)
    - Text that has scrolled offscreen can be reviewed without scrolling the console window. (#12669)
    - More detailed text formatting information is available. (microsoft/terminal#10336)

== Changes for Developers ==
- In builds of Windows Console (`conhost.exe`) with an NVDA API level of 2 (`FORMATTED`) or greater, such as those included with Windows 11 version 22H2 (Sun Valley 2), UI Automation is now used by default. (#10964)
    - This can be overridden by changing the "Windows Console support" setting in NVDA's advanced settings panel.

### CODE REVIEW

**Comment by josephsl** [comment]
Hi, actually you might want to clarify that legacy consoles will be in effect for 20H2 and earlier so Version 2004 can be covered (note that even in Version 2004, newer Windows Terminal preview release may contain UIA console improvements, and checking for 21H1 property says “True”). Thanks.

**Comment by codeofdusk** [comment]
It looks like I've accidentally committed the configobj submodule, how do I fix this?

**Comment by josephsl** [comment]
Hi, if possible, revert the submodule commit, do git submodule update based on master branch, and commit the rest. Thanks.

**Comment by codeofdusk** [comment]
> note that even in Version 2004, newer Windows Terminal preview release may contain UIA console improvements, and checking for 21H1 property says `True`.

How so? That property is on `WinConsoleUIA`, not `WinTerminalUIA`.

**Comment by josephsl** [comment]
Hi, if I do focus.UIATextPattern.GetVisibleRanges().length == 1 from Windows Terminal preview, it returns True (but then again, I might be doing something wrong). Another way to specify 21H1 when we don’t know the build is specifying the first build where the new routine appeared (in this case, 19603, and I can confirm the new console UIA support can be enabled). Thanks.

**Comment by codeofdusk** [comment]
Yes. Since Windows Terminal always has a correct UIA implementation, `ConsoleUIATextInfo` (not `ConsoleUIATextInfoPre21H1`) is always used, and there's no need for an `is21H1Plus` property.

**Comment by feerrenrut** [comment]
@codeofdusk could you update the description to summarize what is rolled back in the two PR's you link to and also summarize the two microsoft/terminal PR's.

**Comment by codeofdusk** [comment]
I've updated the description.

**Comment by ItzLevvie** [comment]
> We don't know if 21H1 will be called version 2103 or 2104. For now, I made it 2103 based on this unofficial page, but we may need to change it.
>
> The 2103 build number in WIN10_VERSIONS_TO_BUILDS will need to be updated once 21H1 goes stable.

Any Windows Insider builds being flighted from the ```RS_PRERELEASE``` branch is no longer tied to any release of Windows since 19536. It's not actually 20H2, or 21H1+, and doesn't have a codename anymore. It now works the same way as how private branches work internally.

20H2 (Manganese) has its own branch now, and 21H1 (Iron) doesn't currently exist.

**Comment by codeofdusk** [comment]
Yes, but any changes made to Open Console before 30 April [are slated for 21H1](https://github.com/microsoft/terminal/milestone/14).

**Comment by codeofdusk** [comment]
The 21H1 deadline has now been extended to 30 July.

PR blocked by:
* [ ] microsoft/terminal#5405 (improve performance of word navigation in some situations)
* [ ] microsoft/terminal#5481 (fix new text diffing in some situations)

Nice to have:
* [ ] microsoft/terminal#2161 (text colours)
* [ ] microsoft/terminal#5406 (create different UIA objects for the standard and alt buffer, obviating #11039)

**Comment by codeofdusk** [comment]
@feerrenrut How can I re-label this PR as Windows Console, not Windows Terminal?

**Comment by feerrenrut** [comment]
Most of the PR seems to be referring to terminal and the microsoft/terminal repo. Can you make the description clearer?

**Comment by codeofdusk** [comment]
Yes, because microsoft/terminal contains both Windows Terminal and Windows Console source code.

**Comment by dingpengyu** [comment]
Will this PR merge NVDA2020.2?
thanks

**Comment by codeofdusk** [comment]
> Will this PR merge NVDA2020.2?
> thanks

Probably not, as there are still [blocking issues](https://github.com/nvaccess/nvda/pull/10964#issuecomment-622344137) on the conhost side. Once the conhost issues are fixed we can look at merging this.

**Comment by codeofdusk** [comment]
FYI deadline has been extended again, it's now 30 September.

**Comment by codeofdusk** [comment]
@feerrenrut @michaelDCurran Could this PR please be labelled as blocked until the blocking PRs are merged?

Also CC @carlos-zamora, @DHowet.

**Comment by leonardder** [comment]
Does this pr also fix the issue where errors are raised when many text flows in at once?

**Comment by codeofdusk** [comment]
This PR makes no functional changes aside from enabling UIA by default.

What API level (NVDA+f1) is your console, and what errors are you seeing?

**Comment by leonardder** [comment]
I opened https://github.com/nvaccess/nvda/issues/12735

**Comment by codeofdusk** [comment]
Looks like that issue is against Windows Terminal, which always uses UIA. This PR is to enable UIA support in conhost.

**Review by Qchristensen — APPROVED** [review]
User Guide looks good, thank you!

**Comment by codeofdusk** [comment]
I found one more issue that, in my opinion, should not block UIA by default as it isn't a regression: microsoft/terminal#11809. This isn't a regression since legacy conhost doesn't seem to support selection *at all* and I can only replicate it under very specific situations.

**Comment by seanbudd** [comment]
This issue was also raised and may block this change: https://github.com/nvaccess/nvda/issues/13318

**Comment by seanbudd** [comment]
Blocked by #13318

**Comment by codeofdusk** [comment]
> Blocked by #13318

No it isn't: this is no different than the previous legacy behaviour.

**Comment by codeofdusk** [comment]
The `FORMATTED` console is now included with Windows 11 Sun Valley 2, available in the external beta channel. I've upgraded the PR to make note of this fact.

**Comment by codeofdusk** [comment]
@seanbudd I'm unsure why the system tests are failing as my changes seem unrelated.

**Review by seanbudd — APPROVED** [review]

**Comment by codeofdusk** [comment]
@seanbudd Thanks for approving! I'm not sure why the system tests are failing as my changes are unrelated.

File: source/UIAHandler/utils.py
Comment (codeofdusk) [reply source/UIAHandler/utils.py:308]: @seanbudd Some of these changes are dependant on one another: the user guide explains the new default behaviour (ternary option, the fact that review commands don't apply to `FORMATTED` consoles), and making the GUI option ternary without changing the meaning of `auto` would mean that `auto` and `legacy` are equivalent (at least making the label of the option confusing if nothing else). This PR should therefore be taken as "all or nothing" (see #10684 for reversion of some of this code when it was decided to not use UIA by default in what are now called `END_EXCLUSIVE` consoles). Note that the option in the config spec has always been ternary: ticking the box sets `winConsoleImplementation=UIA`, and unticking sets `winConsoleImplementation=auto`. (this was done so that user experimentation during the development of the feature would have no bearing on its eventual switch to default).

Comment (seanbudd) [reply source/UIAHandler/utils.py:308]: @codeofdusk - We need a separate PR that just introduces the ternary option in a minimal way.
That way, if we revert the PR, we don't have to revert the ternary option back to a binary checkbox.
It would have been ideal to introduce the option as a ternary option to begin with.

**Comment by codeofdusk** [comment]
> We need a separate PR that just introduces the ternary option in a minimal way.

@seanbudd Done, see #13807 (blocking this PR as it'll probably introduce merge conflicts – I'll handle those once that PR is merged).

**Review by seanbudd — APPROVED** [review]

**Review by Qchristensen — APPROVED** [review]
Reads well, great work!

**Pull request merged**
